### DISCUSSION TITLE
OpenGL API to the Axmol Render Backend API

### DISCUSSION THREAD

**Discussion by thienphuoc** [axmolengine/axmol#1617] [Q&A]
I would like to convert the cool effects from Cocos2d-x 3.17.x to the Axmol engine. However, I'm encountering some issues while converting from the OpenGL API to the Axmol Render Backend API.

For instance, I'm struggling with elements like:

GLenum
glBindBuffer
GLuint
etc.

Could someone assist me in writing a simple guide for this conversion? I aim to learn this technique to convert various effects as needed.

Effect v 3.17.x : 

https://github.com/zelic91/cocos2d-x-effect-nodes/blob/v3/Classes/break/ensBreakNode.cpp

https://github.com/axmolengine/axmol/assets/17909394/a0765e00-3175-46e6-8800-c9db7863cf71

**Comment by rh101** [comment]
Check any OpenGL code in the cpp-tests project of cocos2d-x v3, and then compare that to the cpp-tests in Cocos2d-x v4.  That's where the big changes occurred.  After that, moving to Axmol from Cocos2d-x v4 is straightforward.

I'll try to find any notes I may have made when I initially migrated my projects over to Cocos2d-x v4, and if I do find them, I'll post them here.  Time permitting, I'll see about migrating one of those effects over to Axmol for you so you can see how it's done, but I can't promise that it'll be done any time soon.  Perhaps someone else here has info they can share with you about the migration process.

**Reply by thienphuoc** [reply]
I think it's very necessary, can you do a simple tutorial first, because I have many old libs that need to be converted.

**Reply by rh101** [reply]
As I mentioned, the lack of info isn't actually because of Axmol. The renderer backend API changed between Cocos2d-x v3 and v4, and if I recall correctly, this is the only info they ever provided: https://docs.cocos2d-x.org/cocos2d-x/v4/en/upgradeGuide/

You would need to see an existing Cocos2d-x v3 project that was updated to work with Cocos2d-x v4 to figure out what changes are required, and the only one I know of is the `cpp-tests` project.  So, search for "glBindBuffer" or something similar in the test project of Cocos2d-x v3, and then compare that test code with what was changed in v4, and you'll have your answers.  

Once you know how to port to Cocos2d-x v4, then moving to Axmol requires very few changes (mainly shaders etc.).  Updating to the new shaders in Axmol 2.0 is easy enough, and there are plenty of examples on how it is done in the repo history.

**Comment by aismann** [comment]
Some nice effects! Make a shader is the solution I think.
